Promise/Promise+Extras.swift

```diff
@@ -186,6 +186,12 @@ extension Promise {
             }
         })
     }
+
+    public func mapError(on queue: ExecutionContext = DispatchQueue.main, _ transformError: @escaping (Error) -> Error) -> Promise {
+        return Promise(work: { (fulfill, reject) in
+            self.then(on: queue, fulfill, { reject(transformError($0)) })
+        })
+    }
```

Comment:
I think this would be cleaner if we wrote it in terms of `recover`.

Would something like this work?
```
    public func mapError(on queue: ExecutionContext = DispatchQueue.main, _ transformError: @escaping (Error) -> Error) -> Promise {
         return self.recover({ return Promise(error: transformError($0)) })
     }
```

Reply:
That could work! The tests pass when I make that change too. I have a couple of concerns though: 
* Since `recover` doesn't today take an execution context as a parameter, callers can't control the queue that the transform runs on (and it appears that they always will run on the main thread by my reading of the code). This is easily remedied by providing an execution context parameter to recover though.
* Doing things this way allocates more promises than the way it's written today, which only creates the one being returned. using recover looks like it will allocate four promises during its lifetime if my counting is correct.

I do like the readability of your proposed change though, so if you're OK with those tradeoffs I'm good with making that change.

Reply:
That’s okay with me. This library prioritizes readability over low allocations. Let’s add that execution context too.

Reply:
Cool! all set. I also added a typed variant of mapError to keep things in line with recover/catch, as using recover for this made doing that easy. Let me know if there's anything else you'd like changed :)